mono/mini/interp/transform.c

```diff
@@ -5359,8 +5359,9 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			td->sp = td->stack;
 			SIMPLE_OP (td, MINT_ENDFINALLY);
 			td->last_ins->data [0] = td->clause_indexes [in_offset];
-			// next instructions are always part of new bb
-			td->is_bb_start [td->ip - header->code] = 1;
+			// next instructions, if they exist, are always part of new bb
+			if (td->ip - td->il_code < td->code_size)
+				td->is_bb_start [td->ip - header->code] = 1;
```

Comment:
Could also grow the allocation of `is_bb_start` by one.

Reply:
How does this even happen ? Can a method end with endfinally ? :confused:

If not, I would rather throw invalid IL.

Reply:
I don't know. Let me try to see.

debug_interp also has similar bugs, that make more sense, i.e. during inlining.
```
#0  0x000000000035b720 in generate_code (td=0x1ffeff6ab0, method=0x9c2bb70, header=0xb603530, generic_context=0x9c2bbb0, error=0x1ffeff6d90)
    at interp/transform.c:3102
#1  0x00000000003562f7 in interp_inline_method (td=0x1ffeff6ab0, target_method=0x9c2bb70, header=0xb603530, error=0x1ffeff6d90) at interp/transform.c:1790
#2  0x00000000003578d3 in interp_transform_call (td=0x1ffeff6ab0, method=0xade97f0, target_method=0x9c2bb70, domain=0x6075fc0, generic_context=0xade9830,
    is_bb_start=0xb600970 "\001", constrained_class=0x0, readonly=0, error=0x1ffeff6d90, check_visibility=1, save_last_error=0) at interp/transform.c:2086
#3  0x000000000035dcb6 in generate_code (td=0x1ffeff6ab0, method=0xade97f0, header=0xb601370, generic_context=0xade9830, error=0x1ffeff6d90)
    at interp/transform.c:3371
#4  0x0000000000370db6 in generate (method=0xade97f0, header=0xb601370, rtm=0x1ffeff6c90, generic_context=0xade9830, error=0x1ffeff6d90)
    at interp/transform.c:6376
#

		if (td->verbose_level > 1) {
			g_print ("IL_%04lx %s %-10s, sp %ld, %s %-12s vt_sp %u (max %u)\n",
				td->ip - td->il_code,
				td->is_bb_start [td->ip - td->il_code] == 3 ? "<>" :
```

Reply:
See `TryDequeueSlow`.

Here is a reduced form:

```
class A
{
void a(int b)
{
  while (true)
    lock (this) { }
}
}

csc 3.cs /t:library /optimize

.method private hidebysig instance void  a(int32 b) cil managed
{
  // Code size       24 (0x18)
  .maxstack  2
  .locals init (class A V_0,
           bool V_1)
  IL_0000:  ldarg.0
  IL_0001:  stloc.0
  IL_0002:  ldc.i4.0
  IL_0003:  stloc.1
  .try
  {
    IL_0004:  ldloc.0
    IL_0005:  ldloca.s   V_1
    IL_0007:  call       void [mscorlib]System.Threading.Monitor::Enter(object,
                                                                        bool&)
    IL_000c:  leave.s    IL_0000
  }  // end .try
  finally
  {
    IL_000e:  ldloc.1
    IL_000f:  brfalse.s  IL_0017
    IL_0011:  ldloc.0
    IL_0012:  call       void [mscorlib]System.Threading.Monitor::Exit(object)
    IL_0017:  endfinally
  }  // end handler
} // end of method A::a
```

```

Reply:
I thought ret is always required for method end, I guess not. Make sure the code_size thing. Note that we never inline methods that have clauses.